core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -482,19 +492,29 @@ public void stopPartitions(Set<StopPartition> stopPartitions,
                         task.cancel();
                         return null;
                     });
-                    leaderExpirationRLMTasks.computeIfPresent(tpId, (topicIdPartition, task) -> {
-                        LOGGER.info("Cancelling the expiration RLM task for tpId: {}", tpId);
-                        task.cancel();
-                        return null;
-                    });
                     followerRLMTasks.computeIfPresent(tpId, (topicIdPartition, task) -> {
                         LOGGER.info("Cancelling the follower RLM task for tpId: {}", tpId);
                         task.cancel();
                         return null;
                     });
 
+                    // here, we have to handle retain and delete separately.
+                    // If "retain" is set, we should not cancel expiration task
+                    if (!REMOTE_LOG_DISABLE_POLICY_RETAIN.equals(remoteLogDisablePolicy)) {
+                        leaderExpirationRLMTasks.computeIfPresent(tpId, (topicIdPartition, task) -> {
+                            LOGGER.info("Cancelling the expiration RLM task for tpId: {}", tpId);
+                            task.cancel();
```

Comment:
No, they will get removed because we explicitly delete all the remote log segments in the method [below](https://github.com/apache/kafka/blob/trunk/core/src/main/java/kafka/log/remote/RemoteLogManager.java#L500), not rely on the `leaderExpirationRLMTasks`.

Reply:
I was under the assumption that we will handle KAFKA-15313 together with this patch. 

https://github.com/apache/kafka/pull/13947#discussion_r1281675818

Reply:
We can make delete async in [KAFKA-15313](https://issues.apache.org/jira/browse/KAFKA-15313) for sure. For this PR, I think we just focus on enabling the remote log disable policy feature.